## 1.个人史、疾病史相关性分析：
## 2.血肿/水肿体积相关性分析：
## 3.形状、灰度特征相关性分析：

In [4]:
import pandas as pd
from scipy.stats import spearmanr

# 1. 读取.xlsx文件
data = pd.read_excel('数据.xlsx')

# 定义各类别的特征
history_features = ['高血压病史', '卒中病史', '糖尿病史', '房颤史', '冠心病史', '吸烟史', '饮酒史']
volume_features = ['HM_volume', 'ED_volume', '随访1HM_volume', '随访2HM_volume']
shape_gray_features = [col for col in data.columns if col not in ["患者", "90天mRS"] + history_features + volume_features]

def spearman_correlation_analysis(features):
    correlations = {}
    p_values = {}
    for feature in features:
        corr, p_value = spearmanr(data[feature], data["90天mRS"])
        correlations[feature] = corr
        p_values[feature] = p_value
    return correlations, p_values

# 输出相关性结果为Markdown格式
def output_to_markdown(features, correlations, p_values):
    markdown_output = []
    markdown_output.append("| Feature Name | Spearman Correlation | P-value |")
    markdown_output.append("|--------------|----------------------|---------|")
    for feature in features:
        markdown_output.append(f"| {feature} | {correlations[feature]:.4f} | {p_values[feature]:.4f} |")
    return markdown_output

# 分析各类别特征的相关性
history_corrs, history_pvals = spearman_correlation_analysis(history_features)
volume_corrs, volume_pvals = spearman_correlation_analysis(volume_features)
shape_gray_corrs, shape_gray_pvals = spearman_correlation_analysis(shape_gray_features)

# 输出相关性分析结果
print("\n个人史、疾病史相关性分析：")
for line in output_to_markdown(history_features, history_corrs, history_pvals):
    print(line)

print("\n血肿/水肿体积相关性分析：")
for line in output_to_markdown(volume_features, volume_corrs, volume_pvals):
    print(line)

print("\n形状、灰度特征相关性分析：")
for line in output_to_markdown(shape_gray_features, shape_gray_corrs, shape_gray_pvals):
    print(line)



个人史、疾病史相关性分析：
| Feature Name | Spearman Correlation | P-value |
|--------------|----------------------|---------|
| 高血压病史 | 0.0730 | 0.4703 |
| 卒中病史 | 0.0520 | 0.6076 |
| 糖尿病史 | 0.2995 | 0.0025 |
| 房颤史 | 0.1678 | 0.0953 |
| 冠心病史 | 0.2917 | 0.0032 |
| 吸烟史 | -0.2196 | 0.0282 |
| 饮酒史 | -0.2683 | 0.0070 |

血肿/水肿体积相关性分析：
| Feature Name | Spearman Correlation | P-value |
|--------------|----------------------|---------|
| HM_volume | 0.3664 | 0.0002 |
| ED_volume | 0.2653 | 0.0076 |
| 随访1HM_volume | 0.4320 | 0.0000 |
| 随访2HM_volume | 0.4254 | 0.0000 |

形状、灰度特征相关性分析：
| Feature Name | Spearman Correlation | P-value |
|--------------|----------------------|---------|
| 脑出血前mRS评分 | -0.0165 | 0.8703 |
| 脑室引流 | 0.2050 | 0.0407 |
| 止血治疗 | -0.0561 | 0.5793 |
| 降颅压治疗 | 0.1659 | 0.0991 |
| 降压治疗 | 0.1561 | 0.1208 |
| 镇静、镇痛治疗 | 0.1423 | 0.1580 |
| 止吐护胃 | -0.0804 | 0.4267 |
| 营养神经 | 0.0135 | 0.8943 |
| HM_ACA_R_Ratio | 0.1825 | 0.0692 |
| HM_MCA_R_Ratio | -0.0216 | 0.8307 |
| HM_PCA_R_Ratio | -0.1068 | 

个人史、疾病史相关性分析：
| Feature Name | Spearman Correlation | P-value |
|--------------|----------------------|---------|
| 高血压病史 | 0.0730 | 0.4703 |
| 卒中病史 | 0.0520 | 0.6076 |
| 糖尿病史 | 0.2995 | 0.0025 |
| 房颤史 | 0.1678 | 0.0953 |
| 冠心病史 | 0.2917 | 0.0032 |
| 吸烟史 | -0.2196 | 0.0282 |
| 饮酒史 | -0.2683 | 0.0070 |

血肿/水肿体积相关性分析：
| Feature Name | Spearman Correlation | P-value |
|--------------|----------------------|---------|
| HM_volume | 0.3664 | 0.0002 |
| ED_volume | 0.2653 | 0.0076 |
| 随访1HM_volume | 0.4320 | 0.0000 |
| 随访2HM_volume | 0.4254 | 0.0000 |

形状、灰度特征相关性分析：
| Feature Name | Spearman Correlation | P-value |
|--------------|----------------------|---------|
| 脑出血前mRS评分 | -0.0165 | 0.8703 |
| 脑室引流 | 0.2050 | 0.0407 |
| 止血治疗 | -0.0561 | 0.5793 |
| 降颅压治疗 | 0.1659 | 0.0991 |
| 降压治疗 | 0.1561 | 0.1208 |
| 镇静、镇痛治疗 | 0.1423 | 0.1580 |
| 止吐护胃 | -0.0804 | 0.4267 |
| 营养神经 | 0.0135 | 0.8943 |
| HM_ACA_R_Ratio | 0.1825 | 0.0692 |
| HM_MCA_R_Ratio | -0.0216 | 0.8307 |
| HM_PCA_R_Ratio | -0.1068 | 0.2903 |
| HM_Pons_Medulla_R_Ratio | 0.0557 | 0.5822 |
| HM_Cerebellum_R_Ratio | -0.0826 | 0.4138 |
| HM_ACA_L_Ratio | 0.2639 | 0.0080 |
| HM_MCA_L_Ratio | 0.1765 | 0.0790 |
| HM_PCA_L_Ratio | 0.1606 | 0.1104 |
| HM_Pons_Medulla_L_Ratio | 0.1301 | 0.1970 |
| HM_Cerebellum_L_Ratio | 0.0680 | 0.5014 |
| ED_ACA_R_Ratio | 0.1067 | 0.2905 |
| ED_MCA_R_Ratio | -0.0038 | 0.9704 |
| ED_PCA_R_Ratio | -0.1129 | 0.2635 |
| ED_Pons_Medulla_R_Ratio | 0.0045 | 0.9642 |
| ED_Cerebellum_R_Ratio | -0.1259 | 0.2118 |
| ED_ACA_L_Ratio | 0.1435 | 0.1544 |
| ED_MCA_L_Ratio | 0.0993 | 0.3257 |
| ED_PCA_L_Ratio | 0.0832 | 0.4104 |
| ED_Pons_Medulla_L_Ratio | 0.0089 | 0.9300 |
| ED_Cerebellum_L_Ratio | 0.0554 | 0.5838 |
| HM_ACA_R_Ratio.1 | 0.2548 | 0.0105 |
| HM_MCA_R_Ratio.1 | -0.0226 | 0.8235 |
| HM_PCA_R_Ratio.1 | -0.1044 | 0.3012 |
| HM_Pons_Medulla_R_Ratio.1 | 0.0684 | 0.4986 |
| HM_Cerebellum_R_Ratio.1 | -0.0362 | 0.7204 |
| HM_ACA_L_Ratio.1 | 0.2717 | 0.0063 |
| HM_MCA_L_Ratio.1 | 0.1666 | 0.0975 |
| HM_PCA_L_Ratio.1 | 0.2024 | 0.0435 |
| HM_Pons_Medulla_L_Ratio.1 | 0.1214 | 0.2291 |
| HM_Cerebellum_L_Ratio.1 | 0.1313 | 0.1929 |
| ED_volume.1 | 0.4018 | 0.0000 |
| ED_ACA_R_Ratio.1 | 0.0499 | 0.6219 |
| ED_MCA_R_Ratio.1 | -0.0536 | 0.5960 |
| ED_PCA_R_Ratio.1 | -0.1824 | 0.0694 |
| ED_Pons_Medulla_R_Ratio.1 | 0.0700 | 0.4886 |
| ED_Cerebellum_R_Ratio.1 | -0.1168 | 0.2470 |
| ED_ACA_L_Ratio.1 | 0.1573 | 0.1181 |
| ED_MCA_L_Ratio.1 | 0.1039 | 0.3037 |
| ED_PCA_L_Ratio.1 | 0.1753 | 0.0811 |
| ED_Pons_Medulla_L_Ratio.1 | 0.1056 | 0.2957 |
| ED_Cerebellum_L_Ratio.1 | 0.0148 | 0.8839 |
| HM_ACA_R_Ratio.2 | 0.3046 | 0.0021 |
| HM_MCA_R_Ratio.2 | -0.0045 | 0.9647 |
| HM_PCA_R_Ratio.2 | -0.0600 | 0.5530 |
| HM_Pons_Medulla_R_Ratio.2 | 0.0916 | 0.3648 |
| HM_Cerebellum_R_Ratio.2 | -0.1334 | 0.1859 |
| HM_ACA_L_Ratio.2 | 0.4475 | 0.0000 |
| HM_MCA_L_Ratio.2 | 0.2433 | 0.0147 |
| HM_PCA_L_Ratio.2 | 0.2145 | 0.0321 |
| HM_Pons_Medulla_L_Ratio.2 | 0.1389 | 0.1683 |
| HM_Cerebellum_L_Ratio.2 | -0.0977 | 0.3337 |
| ED_volume.2 | 0.3454 | 0.0004 |
| ED_ACA_R_Ratio.2 | 0.1226 | 0.2245 |
| ED_MCA_R_Ratio.2 | -0.1031 | 0.3075 |
| ED_PCA_R_Ratio.2 | -0.0904 | 0.3712 |
| ED_Pons_Medulla_R_Ratio.2 | 0.1130 | 0.2630 |
| ED_Cerebellum_R_Ratio.2 | -0.1319 | 0.1908 |
| ED_ACA_L_Ratio.2 | 0.3411 | 0.0005 |
| ED_MCA_L_Ratio.2 | 0.1547 | 0.1243 |
| ED_PCA_L_Ratio.2 | 0.2109 | 0.0352 |
| ED_Pons_Medulla_L_Ratio.2 | 0.2232 | 0.0256 |
| ED_Cerebellum_L_Ratio.2 | 0.0011 | 0.9914 |

## 使用Spearman秩相关选择特征

In [8]:
import pandas as pd

data = pd.read_excel('数据.xlsx')

# 计算与'90天mRS'的Spearman秩相关
correlations = data.corr(method='spearman')['90天mRS']

# 选择那些与'90天mRS'有较强相关性的特征
key_features = correlations[correlations.abs() > 0.5].index.tolist()

In [25]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, recall_score, f1_score

# 加载数据
data = pd.read_excel('数据.xlsx')

X = data[key_features]
y = data['90天mRS']

# 将数据分为训练集和测试集
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.7, random_state=156)

# 初始化模型
models = {
    'Logistic Regression': LogisticRegression(),
    'SVM': SVC(),
    'Random Forest': RandomForestClassifier()
}

# 训练并评估每个模型
for model_name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(f'{model_name}:')
    print(f'  Accuracy: {accuracy_score(y_test, y_pred):.2f}')
    print(f'  Recall: {recall_score(y_test, y_pred, average="weighted"):.2f}')
    print(f'  F1 Score: {f1_score(y_test, y_pred, average="weighted"):.2f}')


Logistic Regression:
  Accuracy: 0.81
  Recall: 0.81
  F1 Score: 0.74
SVM:
  Accuracy: 0.96
  Recall: 0.96
  F1 Score: 0.94
Random Forest:
  Accuracy: 1.00
  Recall: 1.00
  F1 Score: 1.00


1. 测试集太小：如果您的数据集不大，并且您的测试集相对较小，那么这种完美的分数可能仅仅是由于随机性。您可以尝试使用交叉验证来得到更稳定的性能估计。
2. 过拟合：虽然这里SVM和随机森林在测试集上的表现非常好，但是如果没有正确地调整参数或者数据集很小，模型可能会过拟合。过拟合的模型在训练数据上的表现很好，但在新的、未见过的数据上可能表现得不太好。

根据上述信息，对血性脑卒中患者预后的影响因素进行总结：

年龄：一般来说，年龄较大的患者更容易受到脑卒中的影响，可能是因为随着年龄的增长，血管的健康状况和其他相关健康问题可能导致更高的风险。

血肿体积：在某些特定的脑部位置，血肿体积与患者的预后关系密切。这可能是因为某些脑部区域对功能更为关键。

水肿增长速度：水肿的快速增长可能会增加脑组织的压迫，导致更严重的伤害。

结合临床知识，为预后评估和临床决策提出以下建议：

早期诊断和评估：考虑到年龄、血肿体积和水肿增长速度等因素，为所有患者进行及时的影像学检查和评估是非常重要的。

高危患者的积极治疗：对于那些存在高风险特征的患者，如大的血肿体积、快速增长的水肿或高龄，建议进行更积极的治疗。这可能包括手术干预、强化的药物治疗或更密集的监护。

多学科团队协作：考虑到脑卒中的复杂性，建议神经学家、神经外科医生、康复医生和其他相关专家组成的多学科团队共同参与患者的治疗和康复。

持续监测和随访：即使在出院后，也建议定期对患者进行随访和检查，以及时发现和处理任何潜在的并发症。